modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/utils.py

```diff
@@ -13,11 +13,220 @@
+        returned.
+      - Non-alphanum parts of the name are encoded, using a customized
+        version of the base64 algorithm, that allows alphanum characters only.
+
+    Parameters
+    ----------
+    name : str, int, float, Timestamp, None, tuple
+        Column name to be encoded.
+    ignore_reserved : bool, default: True
+        Do not encode reserved names.
+
+    Returns
+    -------
+    str
+        Encoded name.
+    """
+    if name is None:
+        return "_N"
+    if isinstance(name, int):
+        return f"_I{str(name)}"
+    if isinstance(name, float):
+        return f"_F{str(name)}"
+    if isinstance(name, Timestamp):
+        return f"_D{encode_col_name((name.timestamp(), str(name.tz)))}"
+    if isinstance(name, tuple):
+        dst = ["_T"]
+        count = len(name)
+        for n in name:
+            dst.append(encode_col_name(n))
+            count -= 1
+            if count != 0:
+                dst.append("_S")  # Separator
+        return "".join(dst)
+    if len(name) == 0:
+        return "_E"
+    if ignore_reserved and (name.startswith(IDX_COL_NAME) or name in _RESERVED_NAMES):
+        return name
+
+    non_alpha = _NON_ALPHANUM_PATTERN.search(name)
+    if not non_alpha:
+        # If the name consists only from alphanum characters, return it as is.
+        return name
+
+    dst = []
+    off = 0
+    while non_alpha:
+        start = non_alpha.start()
+        end = non_alpha.end()
+        dst.append(name[off:start])
+        _quote(name[start:end], dst)
+        off = end
+        non_alpha = _NON_ALPHANUM_PATTERN.search(name, off)
+    dst.append(name[off:])
+    return "".join(dst)
+
+
+@lru_cache(1024)
+def decode_col_name(name: str) -> _COL_NAME_TYPE:
+    """
+    Decode column name, previously encoded with encode_col_name().
+
+    Parameters
+    ----------
+    name : str
+        Encoded name.
+
+    Returns
+    -------
+    str, int, float, Timestamp, None, tuple
+        Decoded name.
+    """
+    if name.startswith("_"):
+        if name.startswith(IDX_COL_NAME) or name in _RESERVED_NAMES:
+            return name
+        char = name[1]
+        if char == "N":
+            return None
+        if char == "I":
+            return int(name[2:])
+        if char == "F":
+            return float(name[2:])
+        if char == "D":
+            stamp = decode_col_name(name[2:])
+            return Timestamp.fromtimestamp(stamp[0], tz=stamp[1])
+        if char == "T":
+            dst = [decode_col_name(n) for n in name[2:].split("_S")]
+            return tuple(dst)
+        if char == "E":
+            return ""
+
+    idx = name.find("_Q")
+    if idx == -1:
+        return name
+
+    dst = []
+    off = 0
+    end = len(name)
+    while idx != -1:
+        dst.append(name[off:idx])
+        off = _unquote(name, dst, idx, end)
+        idx = name.find("_Q", off)
+    dst.append(name[off:])
+    return "".join(dst)
+
+
+def _quote(src: str, dst: List[str]):  # noqa: GL08
```

Comment:
I'm more concerned about the need to maintain this in our code base rather than the performance of this function. 

As we figured out from offline discussion, the custom implementation of the base64 algorithm was introduced in order to ensure that the encoded string will only contain alpha-num characters (excluding '+', '/' and '=' symbols that may appear in the strings encoded by 'base64' python's library), which may be useful if a user wants to interact with HDK via its SQL interface (that only supports alpha-num col names).

Well, I have a few things to disagree with here:
1. If we really want this baseX encoding then we may use 'base32' version from python's standard library. The output alphabet for the 32 version is exactly alpha-num characters, its disadvantage though is that [the result is approximately ~20% longer than with base64](https://en.wikipedia.org/wiki/Base32#Disadvantages), but I think we could put up with it.
2. As far as I understand the only reason for this special strings encoding is that if a column name originally contained an '_S' substring then it breaks the decoding as we reserved the '_S' string as a separator of the tuple values:
    ```python
    >>> multiindex_col_name = ("level1", "level2")
    >>> encode_without_base64(multiindex_col_name)
    "_T_Slevel1_Slevel2"
    >>> decode_col_name("_T_Slevel1_Slevel2")
    ("level1", "level2")
    >>> decode_col_name(encode_without_base64(("lev_Sel1", "level2")))
    ("lev", "el1", "level2")
    ```
    So for me, it feels like an overkill to use baseX encoding algorithm for every string because of that. One of the reasons I'm against it is that such kind of mangling may make debugging really hard as you just wouldn't understand what the column name is right after you imported pandas df to modin:
    ```python
    >>> base64.b32encode(bytes("simple column name", "utf-8"))
    b'ONUW24DMMUQGG33MOVWW4IDOMFWWK==='
    ```
    Maybe we can use some really special separator to split the tuple values in the encoded column names so there would be less chance we met it in a user's column name and then just raise an error if this special character is actually met (just reserve certain symbols and forbid them in column names).

Reply:
> 2\. I'm against it is that such kind of mangling may make debugging really hard as you just wouldn't understand what the column name is right after you imported pandas df to modin

Not, on the modin level the column names are not changed. The encoded names are used in the arrow tables. Another point - alpha-num parts of the string are not encoded. In your example, only the space characters will be encoded.

I agree with you, that maintaining this code could be a pain, despite that the algorithm is quite simple. I've simplified the algorithm and removed the non-alphanum characters encoding.